File: 4.Key-Pair/Javascript/1.js
function pairMatch(pairs) {
    const pairMap = {
        ')': '(',
        '}': '{',
        ']': '['
    };
    const stack = [];

    for (let char of pairs) {
        if (Object.values(pairMap).includes(char)) {
            stack.push(char);
        } else if (pairMap.hasOwnProperty(char)) {
            if (!stack.length || pairMap[char] !== stack.pop()) {
                return false;
            }
        }
    }
    return stack.length === 0;
}

const pairs = "{[()]}";
const res = pairMatch(pairs);
console.log(res); // Output: true

File: 4.Key-Pair/Python/1.py
def pairMatch(pairs):

    pair_map = {")":"(","}":"{","]":"["}      
    stack = []

    for char in pairs:

        if char in pair_map.values():
            stack.append(char)
        elif char in pair_map:
            if stack and pair_map[char] != stack[-1]:
                return False
    return True           
pairs = "{[()]}"
res = pairMatch(pairs)
print(res)
